File: scala-shared/src/main/scala/de/rubixdev/aoc/Vec2.scala
package de.rubixdev.aoc

case class Vec2(x: Long, y: Long) {
  def this(v: Long) = this(v, v)

  def +(other: Vec2): Vec2 = (x + other.x) by (y + other.y)
  def -(other: Vec2): Vec2 = (x - other.x) by (y - other.y)
  def *(scalar: Long): Vec2 = (x * scalar) by (y * scalar)
  def *(scalar: Int): Vec2 = (x * scalar) by (y * scalar)
  def /(scalar: Long): Vec2 = (x / scalar) by (y / scalar)
  def /(scalar: Int): Vec2 = (x / scalar) by (y / scalar)

  def move(direction: Direction): Vec2 = this + direction.vec
}

extension (x: Int) {
  infix def by(y: Int) = Vec2(x.toLong, y.toLong)
  def *(vec: Vec2): Vec2 = vec * x
}

extension (x: Long) {
  infix def by(y: Long) = Vec2(x, y)
  def *(vec: Vec2): Vec2 = vec * x
}

enum Direction:
  case Up, Down, Left, Right
end Direction

extension (dir: Direction) {
  def vec: Vec2 = dir match {
    case Direction.Up    => 0 by -1
    case Direction.Down  => 0 by 1
    case Direction.Left  => -1 by 0
    case Direction.Right => 1 by 0
  }

  def isHorizontal: Boolean = dir match {
    case Direction.Up | Direction.Down    => false
    case Direction.Left | Direction.Right => true
  }

  def rotateLeft: Direction = dir match {
    case Direction.Up    => Direction.Left
    case Direction.Down  => Direction.Right
    case Direction.Left  => Direction.Down
    case Direction.Right => Direction.Up
  }

  def rotateRight: Direction = dir match {
    case Direction.Up    => Direction.Right
    case Direction.Down  => Direction.Left
    case Direction.Left  => Direction.Up
    case Direction.Right => Direction.Down
  }
}

val ADJACENT = Direction.values.map(_.vec) ++ List(
  Direction.Up.vec + Direction.Left.vec,
  Direction.Up.vec + Direction.Right.vec,
  Direction.Down.vec + Direction.Left.vec,
  Direction.Down.vec + Direction.Right.vec,
)

extension [T](map: Seq[Seq[T]]) {
  def apply(index: Vec2): T = map(index.y.toInt)(index.x.toInt)
  def lift2d(index: Vec2): Option[T] =
    map.lift(index.y.toInt).flatMap(_.lift(index.x.toInt))
}

File: kotlin-shared/src/main/kotlin/de/rubixdev/aoc/Vec2.kt
package de.rubixdev.aoc

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

fun List<Int>.toVec2(): Vec2 = this[0] by this[1]

data class Vec2(val x: Long, val y: Long) {
    constructor(v: Long) : this(v, v)

    operator fun plus(other: Vec2) = x + other.x by y + other.y

    operator fun minus(other: Vec2) = x - other.x by y - other.y

    operator fun times(scalar: Long) = x * scalar by y * scalar

    operator fun times(scalar: Int) = x * scalar by y * scalar

    operator fun div(scalar: Long) = x / scalar by y / scalar

    operator fun div(scalar: Int) = x / scalar by y / scalar

    operator fun rem(other: Vec2) = x % other.x by y % other.y

    infix fun mod(other: Vec2) = x.mod(other.x) by y.mod(other.y)

    fun isInBounds(map: Collection<Collection<*>>) = x in map.first().indices && y in map.indices

    fun isInBounds(size: Vec2) = x in 0..size.x && y in 0..size.y

    fun reduce() = gcd(x, y).let { (x / it) by (y / it) }

    fun abs() = abs(x) by abs(y)

    fun area() = x * y
}

operator fun Pair<Vec2, Vec2>.contains(vec: Vec2) =
    vec.x in min(first.x, second.x)..max(first.x, second.x) &&
        vec.y in min(first.y, second.y)..max(first.y, second.y)

operator fun Long.times(vec: Vec2) = vec * this

operator fun Int.times(vec: Vec2) = vec * this

tailrec fun gcd(a: Long, b: Long): Long = if (b == 0L) abs(a) else gcd(b, a % b)

infix fun Int.by(y: Int): Vec2 = Vec2(this.toLong(), y.toLong())

infix fun Long.by(y: Long): Vec2 = Vec2(this, y)

enum class Direction {
    UP,
    DOWN,
    LEFT,
    RIGHT,
    ;

    val vec get() =
        when (this) {
            UP -> 0 by -1
            DOWN -> 0 by 1
            LEFT -> -1 by 0
            RIGHT -> 1 by 0
        }

    val isHorizontal get() =
        when (this) {
            UP, DOWN -> false
            LEFT, RIGHT -> true
        }

    fun rotateLeft() = when (this) {
        UP -> LEFT
        DOWN -> RIGHT
        LEFT -> DOWN
        RIGHT -> UP
    }

    fun rotateRight() = when (this) {
        UP -> RIGHT
        DOWN -> LEFT
        LEFT -> UP
        RIGHT -> DOWN
    }

    fun move(vec2: Vec2) = vec + vec2

    override fun toString() = when (this) {
        UP -> "^"
        DOWN -> "v"
        LEFT -> "<"
        RIGHT -> ">"
    }
}
val ADJACENT = Direction.entries.map { it.vec } + listOf(
    Direction.UP.vec + Direction.LEFT.vec,
    Direction.UP.vec + Direction.RIGHT.vec,
    Direction.DOWN.vec + Direction.LEFT.vec,
    Direction.DOWN.vec + Direction.RIGHT.vec,
)

fun Char.toDirection() = when (this) {
    '^' -> Direction.UP
    '>' -> Direction.RIGHT
    'v' -> Direction.DOWN
    '<' -> Direction.LEFT
    else -> null
}

fun CharSequence.toDirection() = first().toDirection()

operator fun <T> List<List<T>>.get(index: Vec2) = this[index.y.toInt()][index.x.toInt()]

operator fun <T> List<MutableList<T>>.set(index: Vec2, value: T) {
    this[index.y.toInt()][index.x.toInt()] = value
}

fun <T> List<List<T>>.getOrNull(index: Vec2) =
    getOrNull(index.y.toInt())?.getOrNull(index.x.toInt())

fun <T> Iterable<Iterable<T>>.findPos(predicate: (T) -> Boolean) =
    withIndex().firstNotNullOf { (y, line) ->
        line.withIndex().find { (_, tile) -> predicate(tile) }?.let { it.index by y }
    }
